lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	wg     sync.WaitGroup
+
+	replayStore replayStore
+
+	// runCtx is a context that gets canceled when all connections should be
+	// ungracefully terminated.
+	runCtx context.Context
+	// runCancel cancels runCtx.
+	runCancel context.CancelFunc
+	// serveCtx is a context that gets canceled when all listeners should stop
+	// accepting new connections.
+	serveCtx context.Context
+	// serveCancel cancels serveCtx.
+	serveCancel context.CancelFunc
+}
 
 // NewQUICServer returns a [QUICServer] with the given config.
 func NewQUICServer(cfg QUICServerConfig) (*QUICServer, error) {
 	if err := cfg.checkAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	panic("QUIC proxy peering is not implemented")
+
+	tlsConfig := utils.TLSConfig(cfg.CipherSuites)
+	tlsConfig.GetCertificate = cfg.GetCertificate
+	tlsConfig.VerifyPeerCertificate = verifyPeerCertificateIsProxy
+	tlsConfig.NextProtos = []string{quicNextProto}
+	tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	tlsConfig.MinVersion = tls.VersionTLS13
+
+	getClientCAs := cfg.GetClientCAs
+	tlsConfig.GetConfigForClient = func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+		clientCAs, err := getClientCAs(chi)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		utils.RefreshTLSConfigTickets(tlsConfig)
+		c := tlsConfig.Clone()
+		c.ClientCAs = clientCAs
+		return c, nil
+	}
+
+	quicConfig := &quic.Config{
+		MaxStreamReceiveWindow:     quicMaxReceiveWindow,
+		MaxConnectionReceiveWindow: quicMaxReceiveWindow,
+
+		MaxIncomingStreams:    1 << 60,
+		MaxIncomingUniStreams: -1,
+
+		MaxIdleTimeout:  quicMaxIdleTimeout,
+		KeepAlivePeriod: quicKeepAlivePeriod,
+
+		Allow0RTT: true,
+	}
+
+	runCtx, runCancel := context.WithCancel(context.Background())
+	serveCtx, serveCancel := context.WithCancel(runCtx)
+
+	return &QUICServer{
+		log:           cfg.Log,
+		clusterDialer: cfg.ClusterDialer,
+		tlsConfig:     tlsConfig,
+		quicConfig:    quicConfig,
+
+		runCtx:      runCtx,
+		runCancel:   runCancel,
+		serveCtx:    serveCtx,
+		serveCancel: serveCancel,
+	}, nil
 }
 
 // Serve opens a listener and serves incoming connection. Returns after calling
 // Close or Shutdown.
 func (s *QUICServer) Serve(t *quic.Transport) error {
-	panic("QUIC proxy peering is not implemented")
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return trace.Wrap(quic.ErrServerClosed)
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+	s.mu.Unlock()
+
+	lis, err := t.ListenEarly(s.tlsConfig, s.quicConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer lis.Close()
+	defer context.AfterFunc(s.serveCtx, func() { _ = lis.Close() })()
+
+	for {
+		// the listener will be closed when serveCtx is done, but Accept will
+		// return any queued connection before erroring out with a
+		// [quic.ErrServerClosed]
+		c, err := lis.Accept(context.Background())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		s.wg.Add(1)
+		go s.handleConn(c)
+	}
+}
+
+func (s *QUICServer) handleConn(c quic.EarlyConnection) {
```

Comment:
nit: Pass the context down from Serve?

```suggestion
func (s *QUICServer) handleConn(ctx context.Context, conn quic.EarlyConnection) {
```

Reply:
Nothing but the `AfterFunc(runCtx)` in there should be using a context other than the connection's context anyway - I'll update the function to do that.